compiler/src/dmd/dtemplate.d

```diff
@@ -6293,6 +6300,8 @@ extern (C++) class TemplateInstance : ScopeDsymbol
                     minst = tnext.minst; // cache result from non-speculative sibling
                     return false;
                 }
```

Comment:
I think these (incl. other nested function) might need work - you're now terminating the siblings iteration on the first non-speculative sibling if the primary one is speculative. The previous recursive `needsCodegen()` call meant that `tnext.minst` had propagated all the way to the left from the first sibling that needed codegen (=> potentially iterating further).

There's https://github.com/dlang/dmd/blob/ee1acbb62dc4a65e72d7bf27e61761fb34d497e2/compiler/src/dmd/dsymbolsem.d#L5941-L5953 (overriding a primary instance's `minst` if it was speculative so far but got a non-speculative sibling from a root module), but I don't think that covers all cases (and depending on that detail in another source file certainly not ideal).

Reply:
Getting rid of the `return false;` should do here - if the primary one is speculative, overwrite it to first non-speculative sibling's module. Then continue the iteration, looking for a root sibling - and overwrite it to that module then, and terminate iteration.